### ISSUE TITLE
[input chunk] fs_chunk_size is updated by 0 bytes and will cause "no available chunk" error

### ISSUE THREAD

**Issue by JeffLuoo** [fluent/fluent-bit#3094]
## Bug Report
Saw lots of logs generated by fluent bit that the fs_chunk_size is only updated by 0 byte
```
[2021/02/19 01:33:38] [debug] [storage] [cio file] alloc_size from 4096 to 36864
[2021/02/19 01:33:38] [debug] [input chunk] update plugin forward.1 fs_chunk_size by 0 bytes, the current fs_chunk_size is 80118618 bytes
[2021/02/19 01:33:38] [debug] [storage] [cio file] synced at: tail.0/1-1613698418.145542736.flb
[2021/02/19 01:33:38] [debug] [storage] tail.0:1-1613698418.146209340.flb mapped OK
[2021/02/19 01:33:38] [debug] [storage] [cio file] synced at: tail.0/1-1613698418.146209340.flb
[2021/02/19 01:33:38] [debug] [input chunk] chunk 1-1613698418.146209340.flb required 1185 bytes and 1919881547 bytes left in plugin forward.1
[2021/02/19 01:33:38] [debug] [storage] tail.0:1-1613698418.146209340.flb mapped OK
[2021/02/19 01:33:38] [debug] [storage] [cio file] alloc_size from 4096 to 36864
[2021/02/19 01:33:38] [debug] [input chunk] update plugin forward.1 fs_chunk_size by 0 bytes, the current fs_chunk_size is 80118453 bytes
[2021/02/19 01:33:38] [debug] [storage] [cio file] synced at: tail.0/1-1613698418.146209340.flb
[2021/02/19 01:33:38] [debug] [input:tail:tail.0] inode=6451885 events: IN_MODIFY
[2021/02/19 01:33:38] [debug] [storage] tail.0:1-1613698418.146753342.flb mapped OK
[2021/02/19 01:33:38] [debug] [storage] [cio file] synced at: tail.0/1-1613698418.146753342.flb
[2021/02/19 01:33:38] [debug] [input chunk] chunk 1-1613698418.146753342.flb required 1197 bytes and 1919881712 bytes left in plugin forward.1
[2021/02/19 01:33:38] [debug] [storage] tail.0:1-1613698418.146753342.flb mapped OK
[2021/02/19 01:33:38] [debug] [storage] [cio file] alloc_size from 4096 to 36864
[2021/02/19 01:33:38] [debug] [input chunk] update plugin forward.1 fs_chunk_size by 0 bytes, the current fs_chunk_size is 80118288 bytes
[2021/02/19 01:33:38] [debug] [storage] [cio file] synced at: tail.0/1-1613698418.146753342.flb
[2021/02/19 01:33:38] [debug] [input:tail:tail.0] inode=6451885 events: IN_MODIFY
[2021/02/19 01:33:38] [debug] [storage] tail.0:1-1613698418.147251245.flb mapped OK
```

And this will cause the "no available chunk" issue even if there are still enough space (I guess it is caused by fs_chunk_size is not updated correctly)
```
[2021/02/19 00:54:52] [error] [input chunk] no available chunk
[2021/02/19 00:54:52] [error] [input chunk] no enough space in filesystem to buffer chunk 1-1613696092.707366780.flb in plugin forward.1
[2021/02/19 00:54:52] [error] [input chunk] no available chunk
[2021/02/19 00:54:52] [error] [input chunk] no enough space in filesystem to buffer chunk 1-1613696092.707528581.flb in plugin forward.1
[2021/02/19 00:54:52] [error] [input chunk] no available chunk
[2021/02/19 00:54:52] [error] [input chunk] no enough space in filesystem to buffer chunk 1-1613696092.808458488.flb in plugin forward.0
[2021/02/19 00:54:52] [error] [input chunk] no available chunk
[2021/02/19 00:54:52] [error] [input:emitter:rewrite_foo] error registering chunk with tag: foo
[2021/02/19 00:54:52] [error] [input chunk] no enough space in filesystem to buffer chunk 1-1613696092.858187237.flb in plugin forward.0
```

**Describe the bug**

Error `no available chunk` even if there are enough spaces left.

**To Reproduce**
- Steps to reproduce the problem:
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder
data:
  fluent-bit.conf: |
    [SERVICE]
        Flush                     1
        Log_Level                 debug
        Parsers_File              /fluent-bit/etc/parsers.conf
        Plugins_File              /fluent-bit/etc/plugins.conf
        HTTP_Server               On
        storage.path              /var/log/flb-storage/
        storage.max_chunks_up     64
        storage.backlog.mem_limit 128M
        storage.metrics           on
    
    [INPUT]
        Name              tail
        Tag               kubernetes.*
        Path              /var/log/containers/*.log
        Parser            docker
        Docker_Mode       On
        DB                /var/log/flb-tail.db
        DB.sync           normal
        DB.locking        true
        Refresh_Interval  15
        Read_from_Head    On
        Buffer_Chunk_Size 128K
        Buffer_Max_Size   128K
        Skip_Long_Lines   On
        Mem_Buf_Limit     30M
        storage.type      filesystem

    [FILTER]
        Name                kubernetes
        Match               kubernetes.var.log.containers.*
        Kube_Tag_Prefix     kubernetes.var.log.containers.
        Annotations         Off
        K8S-Logging.Exclude On

    [FILTER]
        Name                  rewrite_tag
        Match                 kubernetes.var.log.containers.*
        Rule                  $kubernetes['namespace_name'] ^(.*)$ kubernetes.$1 false
        Emitter_Name          rewrite
        Emitter_Storage.type  filesystem
        Emitter_Mem_Buf_Limit 30M

    [FILTER]
        Name                  rewrite_tag
        Match                 kubernetes.logging
        Rule                  $kubernetes['namespace_name'] ^(.*)$ foo false
        Emitter_Name          rewrite_foo
        Emitter_Storage.type  filesystem
        Emitter_Mem_Buf_Limit 30M

    [OUTPUT]
        Name                     forward
        Match                    foo
        Host                     fluentd.logging
        Port                     24224
        Tag                      foo
        Retry_Limit              false
        storage.total_limit_size 100M
    
    [OUTPUT]
        Name                     forward
        Match                    kubernetes.*
        Host                     fluentd.logging
        Port                     24224
        Tag                      bar
        Retry_Limit              false
        storage.total_limit_size 100M
---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder
rules:
  - apiGroups:
      - ""
    resources:
      - "namespaces"
      - "services"
      - "pods"
    verbs:
      - "get"
      - "watch"
      - "list"

---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder
roleRef:
  kind: ClusterRole
  apiGroup: rbac.authorization.k8s.io
  name: fluentbit-forwarder
subjects:
  - kind: ServiceAccount
    name: fluentbit-forwarder
    namespace: logging

---
apiVersion: v1
kind: Service
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder
    app: fluentbit-forwarder
spec:
  type: ClusterIP
  ports:
    - name: http
      port: 2020
      protocol: TCP
      targetPort: http
  selector:
    app.kubernetes.io/name: fluentbit-forwarder

---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: fluentbit-forwarder
  labels:
    app.kubernetes.io/name: fluentbit-forwarder
    app: fluentbit-forwarder
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: fluentbit-forwarder
  updateStrategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        app.kubernetes.io/name: fluentbit-forwarder
        app: fluentbit-forwarder
    spec:
      serviceAccountName: fluentbit-forwarder
      securityContext:
        runAsGroup: 65532
        runAsUser: 65532
      containers:
        - name: fluentbit
          image: fluent/fluent-bit:1.7-debug
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 2020
              name: http
              protocol: TCP
          resources:
            limits:
              cpu: 100m
              memory: 200Mi
            requests:
              cpu: 100m
              memory: 200Mi
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
              - ALL
            privileged: false
            runAsGroup: 0
            runAsUser: 0
          volumeMounts:
            - name: varlog
              mountPath: /var/log
            - name: varlibdockercontainers
              mountPath: /var/lib/docker/containers
              readOnly: true
            - mountPath: /fluent-bit/etc/fluent-bit.conf
              name: config
              subPath: fluent-bit.conf
      volumes:
        - name: varlog
          hostPath:
            path: /var/log
        - name: varlibdockercontainers
          hostPath:
            path: /var/lib/docker/containers
        - configMap:
            name: fluentbit-forwarder
            items:
            - key: fluent-bit.conf
              path: fluent-bit.conf
          name: config
```
**Expected behavior**

The `fs_chunk_size` is updated correctly.
**Screenshots**

**Your Environment**

* Version used: fluent/fluent-bit:1.7-debug
* Configuration:
* Environment name and version (e.g. Kubernetes? What version?):
* Server type and version:
* Operating System and version:
* Filters and plugins:

**Additional context**

**Comment by JeffLuoo** [comment]
Some Notes:
During testing, I found that the fs_chunks_size for `forward.0` is updated correctly and the error only happens on `forward.1`.

Chatted with @panaji and he confirmed that this error happens in v1.7 as he tested it on v1.6.10 and didn't see this error.

**Comment by JeffLuoo** [comment]
can be reproduced locally:
```
[SERVICE]
    Flush                     1
    Log_Level                 debug
    Parsers_File              /fluent-bit/etc/parsers.conf
    HTTP_Server               On
    storage.path              /var/log/flb-storage/
    storage.max_chunks_up     64
    storage.backlog.mem_limit 30M
    storage.metrics           on

[INPUT]
    Name              dummy
    Tag               dummy.first
    Dummy             {"tool": "fluent Dummy 1", "sub": {"s1": {"s2": "bittt1"}}}
    Parser            json
    Mem_Buf_Limit     30M
    storage.type      filesystem

[INPUT]
    Name              dummy
    Tag               dummy.second
    Dummy             {"tool": "fluent Dummy 2", "sub": {"s1": {"s2": "bittt2"}}}
    Parser            json
    Mem_Buf_Limit     30M
    storage.type      filesystem

[FILTER]
    Name                  rewrite_tag
    Match                 dummy.first
    Rule                  $tool ^(.*)$ kubernetes.test false
    Emitter_Name          rewrite
    Emitter_Storage.type  filesystem
    Emitter_Mem_Buf_Limit 30M

[FILTER]
    Name                  rewrite_tag
    Match                 kubernetes.logging
    Rule                  $tool ^(.*)$ foo false
    Emitter_Name          rewrite_foo
    Emitter_Storage.type  filesystem
    Emitter_Mem_Buf_Limit 30M

[OUTPUT]
    Name                     stdout
    Match                    foo
    Retry_Limit              false
    storage.total_limit_size 50M

[OUTPUT]
    Name                     stdout
    Match                    kubernetes.*
    Retry_Limit              false
    storage.total_limit_size 50M
```

**Comment by JeffLuoo** [comment]
Beside from the above issue, I also found a bug that `fs_chunk_size` is not updated correctly...
This bug can be reproduced locally by:
```
[SERVICE]
    Flush                     1
    Log_Level                 debug
    Parsers_File              /fluent-bit/etc/parsers.conf
    HTTP_Server               On
    storage.path              /var/log/flb-storage/
    storage.max_chunks_up     64
    storage.backlog.mem_limit 30M
    storage.metrics           on

[INPUT]
    Name              dummy
    Tag               dummy.first
    Dummy             {"tool": "fluent Dummy 1", "sub": {"s1": {"s2": "bittt1"}}}
    Parser            json
    Mem_Buf_Limit     30M
    storage.type      filesystem

[FILTER]
    Name                  rewrite_tag
    Match                 dummy.first
    Rule                  $tool ^(.*)$ kubernetes.test false
    Emitter_Name          rewrite
    Emitter_Storage.type  filesystem
    Emitter_Mem_Buf_Limit 30M

[OUTPUT]
    Name                     stdout
    Match                    kubernetes.*
    Retry_Limit              false
    storage.total_limit_size 50M
```

And from the log I saw
```
[2021/02/21 17:50:35] [debug] [storage] dummy.0:2725659-1613929835.690724872.flb mapped OK
[2021/02/21 17:50:35] [debug] [storage] [cio file] synced at: dummy.0/2725659-1613929835.690724872.flb
[2021/02/21 17:50:35] [debug] [storage] rewrite:2725659-1613929835.691139935.flb mapped OK
[2021/02/21 17:50:35] [debug] [input chunk] chunk 2725659-1613929835.691139935.flb required 51 bytes and 50000000 bytes left in plugin stdout.0
[2021/02/21 17:50:35] [debug] [input chunk] update plugin stdout.0 fs_chunk_size by 51 bytes, the current fs_chunk_size is 51 bytes
[2021/02/21 17:50:36] [debug] [storage] [cio file] synced at: rewrite/2725659-1613929835.691139935.flb
[0] kubernetes.test: [1613929835.690717105, {"tool"=>"fluent Dummy 1", "sub"=>{"s1"=>{"s2"=>"bittt1"}}}]
[2021/02/21 17:50:36] [debug] [task] created task=0x7fe988015620 id=0 OK
[2021/02/21 17:50:36] [debug] [storage] dummy.0:2725659-1613929836.690889846.flb mapped OK
[2021/02/21 17:50:36] [debug] [storage] [cio file] synced at: dummy.0/2725659-1613929836.690889846.flb
[2021/02/21 17:50:36] [debug] [storage] rewrite:2725659-1613929836.691273508.flb mapped OK
[2021/02/21 17:50:36] [debug] [input chunk] chunk 2725659-1613929836.691273508.flb required 51 bytes and 49999949 bytes left in plugin stdout.0
[2021/02/21 17:50:36] [debug] [input chunk] update plugin stdout.0 fs_chunk_size by 51 bytes, the current fs_chunk_size is 102 bytes
[2021/02/21 17:50:36] [debug] [out coro] cb_destroy coro_id=0
[2021/02/21 17:50:36] [debug] [task] destroy task=0x7fe988015620 (task_id=0)
[2021/02/21 17:50:36] [debug] [input chunk] Delete chunk 2725659-1613929835.691139935.flb from plugin stdout.0 fs_chunk_size by 90 bytes, the current fs_chunk_size is 102 bytes
[0] kubernetes.test: [1613929836.690885651, {"tool"=>"fluent Dummy 1", "sub"=>{"s1"=>{"s2"=>"bittt1"}}}]
[2021/02/21 17:50:37] [debug] [storage] [cio file] synced at: rewrite/2725659-1613929836.691273508.flb

```

**Comment by JeffLuoo** [comment]
I printed out the content size vs real size when fluent bit deletes the chunk and found that 
```
[2021/02/21 18:05:43] [debug] [input chunk] Upon delete, real size 90 vs content size 51
```

So my guess is that when data comes in, `fs_chunk_size` is firstly updated by how many bytes of data ingested and the size is calculate the 'real' new bytes being added after the filtering phase. When chunk is destroyed, the chunk contains some meta data which will increase the chunk size which cause the issue.

**Comment by JeffLuoo** [comment]
closed as fixed by #3118
